hashicorp/terraform-provider-tfe
# Add Workspace Outputs data source

## Description

This PR introduces a new data source for handling state outputs. `tfe_outputs ` takes an organization and workspace, and retrieves the state outputs for that workspace. 

**Example**

```
data "tfe_outputs" "foobar" {
  organization = "<org-name>"
  workspace = "<workspace-name>"
}

output "states" {
	value = data.tfe_outputs.foobar.values.<key>
}
```

The `<key>` in the `states` output can be any field in your state outputs. So if your state output looks something like this:
```
  "outputs": {
    "user": {
      "value": "9023256633839603543",
      "type": "string"
    },
    "records": {
      "value": ["hashicorp.com","terraform.io"],
      "type": ["list","string"]
    },
    "secret": {
      "value": "token",
      "type": "string",
      "sensitive": true,
    }
  }
```

Then the `<key>` can be as follows:
```
data "tfe_outputs" "foobar" {
  organization = "<org-name>"
  workspace = "<workspace-name>"
}

output "user_output" {
	value = data.tfe_outputs.foobar.values.user
}
output "records_output" {
	value = data.tfe_outputs.foobar.values.records
}
```

To show sensitive values, the user can set the boolean flag `sensitive=true`:

```
data "tfe_outputs" "foobar" {
  organization = "<org-name>"
  workspace = "<workspace-name>"
  show_sensitive = true
}

output "secret" {
	value = data. tfe_outputs.foobar.values.secret
}
```


## Testing plan

1. Pull Down the branch
2. Create a terraform configuration file that includes the tfe-provider
3. Copy the example code above and make sure the `tfe_outputs ` data source has a organization and workspace that has state outputs. If the workspace doesn't, then the `values` field of the `tfe_outputs ` will be empty.
4. Run terraform and you will see the outputs from that statefile. 
5. Change your terraform config to select a specific value from the output. Re-run terraform and you should see the value printed.

## External links

An RFC was passed internally. 

## Output from acceptance tests


```
$ TFE_TOKEN=<token> TFE_HOSTNAME=app.terraform.io TF_ACC=1 TF_LOG_PROVIDER=DEBUG  go test ./tfe/... -v -run TestAccTFEStateOutputs
=== RUN   TestAccTFEOutputs
--- PASS: TestAccTFEOutputs (12.08s)
=== RUN   TestAccTFEOutputs_emptyOutputs
--- PASS: TestAccTFEOutputs_emptyOutputs (12.12s)
PASS
```
